sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala

```diff
@@ -304,24 +337,40 @@ abstract class HostCoalesceIteratorBase[T <: AutoCloseable : ClassTag](
 
   private def concatenateTablesInHost(): CoalescedHostResult = {
     val result = withResource(new MetricRange(concatTimeMetric)) { _ =>
-      val input = for (_ <- 0 until numTablesInBatch) yield serializedTables.removeFirst()
+      val input = new ArrayBuffer[T]()
+      for (_ <- 0 until numTablesInBatch)
+        input += serializedTables.removeFirst()
+
+      {
+        // update the stats for the next batch in progress
+        numTablesInBatch = serializedTables.size
```

Comment:
These need to be protected by a lock, since two threads may modify these variables at the same time.

Reply:
which two threads do you mean ?

Reply:
The task main thread and the async background thread form the thread pool

Reply:
make sense. I'll rework on thread safety

Reply:
offline synced with @firestarman ,  the code is actualy thread safe. I added more comments